### PR TITLE
feat: switch printing from hex to base64

### LINKED ISSUE #91: Don't use hex for display purposes
We use hex to display several binary things: hashes, auth token, peer ID, ticket.  This is pretty verbose and we can probably use something better.

### LINKED ISSUE #728: Don't use hex for display purposes
We use hex to display several binary things: hashes, auth token, peer ID, ticket.  This is pretty verbose and we can probably use something better.

### PR DESCRIPTION
**Pull request by dignifiedquire**
Why base64? 
Because most of the world uses it already, and the url safe version can be at least easily used in URLs. Also it is very fast :) 

Screenshot 2023-02-02 at 18 27 41

Closes n0-computer/iroh#728

### CODE REVIEW

File: src/bao_slice_decoder.rs [hunk 1]
```diff
@@ -806,8 +808,8 @@ mod tests {
                     let data = &slice[offset..offset + 64];
                     let used = |b| if b { "*" } else { " " };
                     println!("hashes root={is_root}");
-                    println!("{} {}", hex::encode(&data[..32]), used(left));
-                    println!("{} {}", hex::encode(&data[32..]), used(right));
+                    println!("{} {}", util::encode(&data[..32]), used(left));
```
Comment (rklaehn) [inline src/bao_slice_decoder.rs]: I think for the utils for the slice extractor you really want hex.

This is in test, so we can just have hex as a test dependency.

File: src/bao_slice_decoder.rs [hunk 2]
```diff
@@ -818,7 +820,7 @@ mod tests {
                     let data = &slice[offset..offset + size as usize];
                     println!("data range={}..{} root={}", start, start + size, is_root);
                     for chunk in data.chunks(32) {
-                        println!("{}", hex::encode(chunk));
+                        println!("{}", util::encode(chunk));
```
Comment (rklaehn) [inline src/bao_slice_decoder.rs]: same

**Review by rklaehn — APPROVED** [review]
Works for me, but I would prefer if we would use base64 just for the combined token that is going to end up on the QR code.

For a 32 byte hash or peer id, I think hex is fine.

For the debug tools for AsyncSliceDecoder, it *has* to be hex though. (That thing will vanish from the crate as long as we make progress with bao one way or another.. )

File: src/bao_slice_decoder.rs (hunk 1 above)
Comment (dignifiedquire) [reply src/bao_slice_decoder.rs]: sure, works for me

**Comment by dignifiedquire** [comment]
> For a 32 byte hash or peer id, I think hex is fine.

I agree in general, but I am really not a big fan in mixing encodings in user facing pieces.

File: src/bao_slice_decoder.rs (hunk 2 above)
Comment (dignifiedquire) [reply src/bao_slice_decoder.rs]: fixed

File: src/bao_slice_decoder.rs (hunk 1 above)
Comment (dignifiedquire) [reply src/bao_slice_decoder.rs]: fixed

**Comment by rklaehn** [comment]
> > For a 32 byte hash or peer id, I think hex is fine.
> 
> I agree in general, but I am really not a big fan in mixing encodings in user facing pieces.

As long as I don't get weird base64 in my bao format dumper, I am fine with that...

File: Cargo.toml [hunk 3]
```diff
@@ -8,20 +8,19 @@ authors = ["dignifiedquire <me@dignifiedquire.com>"]
 repository = "https://github.com/n0-computer/sendme"
 
 # Sadly this also needs to be updated in .github/workflows/ci.yml
-rust-version = "1.63"
+rust-version = "1.64"
```
Comment (flub) [inline Cargo.toml:11]: Hum, that's unfortunate as it brings us out of sync with deltachat.  Can this not be avoided?  IIRC the `data-encoding` crate is also popular (and some argue better maintained).

Comment (dignifiedquire) [reply Cargo.toml:11]: this has nothing to do with base64, this is coming from `clap`

Comment (dignifiedquire) [reply Cargo.toml:11]: (I ran `cargo update`..)

Comment (flub) [reply Cargo.toml:11]: Is it that important to have the latest clap in the lock file?  I'm not sure that's worth the tradeoff.  For a library we'll have to get used to providing some kind of msrv for a while.

Comment (dignifiedquire) [reply Cargo.toml:11]: No, but it will happen sooner or later, having to pin to an old version of clap is quite unfortunate.

**Comment by flub** [comment]
(don't consider this PR blocked because of my msrv musings btw - not sure how important they are)

**Pull request merged**